### PR TITLE
Snowbridge free consensus updates

### PR DESCRIPTION
**Pull request by claravanstaden**
Allow free Snowbridge consensus updates, if the header interval is larger than the configured value (set to 32, so once a epoch).

This PR also moves the Rococo Snowbridge pallet config into its own module.

Original PR: https://github.com/Snowfork/polkadot-sdk/pull/159

### CODE REVIEW

File: bridges/snowbridge/pallets/ethereum-client/src/lib.rs [hunk 1]
```diff
@@ -634,11 +645,29 @@ pub mod pallet {
...
+		/// successful sync committee updates are free.
+		pub(super) fn check_refundable(update: &Update, latest_slot: u64) -> Pays {
+			// If the sync committee was successfully updated, the update may be free.
+			if update.next_sync_committee_update.is_some() {
+				return Pays::No;
+			}
+
+			// If free headers are allowed and the latest finalized header is larger than the
+			// minimum slot interval, the header import transaction is free.
+			if update.finalized_header.slot >= latest_slot + T::FreeHeadersInterval::get() as u64 {
```
Comment (vgeddes) [inline bridges/snowbridge/pallets/ethereum-client/src/lib.rs:666]: Should use saturating arithmetic to prevent panic.

Comment (claravanstaden) [reply bridges/snowbridge/pallets/ethereum-client/src/lib.rs]: Added in https://github.com/paritytech/polkadot-sdk/pull/5201/commits/9391a36368e962f05a1987bff7e9dda82a54b9cf.

File: bridges/snowbridge/pallets/ethereum-client/src/lib.rs [hunk 2]
```diff
@@ -204,11 +210,10 @@ pub mod pallet {
 		#[transactional]
```
Comment (yrong) [inline bridges/snowbridge/pallets/ethereum-client/src/lib.rs:210]: Is the transactional macro here required?

**Review by vgeddes — APPROVED** [review]

**Review by yrong — APPROVED** [review]
+1

File: bridges/snowbridge/pallets/ethereum-client/src/lib.rs (hunk 2 above)
Comment (claravanstaden) [reply bridges/snowbridge/pallets/ethereum-client/src/lib.rs:210]: Probably not, will clean up later. :)

File: bridges/snowbridge/pallets/ethereum-client/src/tests.rs [hunk 3]
```diff
@@ -129,6 +128,23 @@ pub fn compute_domain_bls() {
...
 }
 
+#[test]
+pub fn may_refund_call_fee() {
+	let finalized_update = Box::new(load_next_finalized_header_update_fixture());
+	let sync_committee_update = Box::new(load_sync_committee_update_fixture());
+	new_tester().execute_with(|| {
+		// Not free, smaller than the allowed free header interval
+		assert_eq!(
+			EthereumBeaconClient::check_refundable(&finalized_update.clone(), 8190),
```
Comment (franciscoaguirre) [inline bridges/snowbridge/pallets/ethereum-client/src/tests.rs:138]: It might be clearer if you use `finalized_update.finalized_header.slot + some_offset`

File: cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/bridge_to_ethereum_config.rs [hunk 4]
```diff
@@ -25,3 +48,173 @@ pub type SnowbridgeExporter = EthereumBlobExporter<
...
+			epoch: 132608,
+		},
+	};
+}
+
+impl snowbridge_pallet_ethereum_client::Config for Runtime {
+	type RuntimeEvent = RuntimeEvent;
+	type ForkVersions = ChainForkVersions;
+	// Free consensus update every epoch. Works out to be 225 updates per day.
+	type FreeHeadersInterval = ConstU32<32>;
```
Comment (franciscoaguirre) [inline cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/bridge_to_ethereum_config.rs:166]: I guess the epoch will not change often, but probably better to use a constant like `EPOCH_LENGTH`

File: prdoc/pr_5201.prdoc [hunk 5]
```diff
@@ -0,0 +1,20 @@
+# Schema: Polkadot SDK PRDoc Schema (prdoc) v1.0.0
+# See doc at https://raw.githubusercontent.com/paritytech/polkadot-sdk/master/prdoc/schema_user.json
+
+title: Snowbridge free consensus updates
+
+doc:
+  - audience: Runtime Dev
+    description: |
+      Allow free consensus updates to the Snowbridge Ethereum client if the headers are more than a certain
+      number of headers apart.
```
Comment (franciscoaguirre) [inline prdoc/pr_5201.prdoc:10]: How does this impact the devs? Is it cheaper to run relayers now?
Does this affect the user somehow? E.g. cheaper to use the bridge?

**Review by serban300 — APPROVED** [review]

File: bridges/snowbridge/pallets/ethereum-client/src/lib.rs [hunk 6]
```diff
@@ -465,11 +470,17 @@ pub mod pallet {
...
+				None => T::WeightInfo::submit(),
+				Some(_) => T::WeightInfo::submit_with_sync_committee(),
+			};
+
 			if update.finalized_header.slot > latest_finalized_state.slot {
 				Self::store_finalized_header(update.finalized_header, update.block_roots_root)?;
 			}
 
-			Ok(())
+			Ok(PostDispatchInfo { actual_weight: Some(actual_weight), pays_fee })
```
Comment (acatangiu) [inline bridges/snowbridge/pallets/ethereum-client/src/lib.rs:483]: Can you double-check and confirm that this doesn't allow free spamming for this call with using the same "valid" update (with `update.next_sync_committee_update.is_some() == true`) over and over for free, potentially DoS-ing BH blocks?

It looks like there is [a `compute_period` verification](https://github.com/paritytech/polkadot-sdk/blob/6ad199baa81035bbeb3f3859c1ad730f08b1a04a/bridges/snowbridge/pallets/ethereum-client/src/lib.rs#L313C1-L322C5) for updating sync committee, that would not allow the DoS I described above, but I got a bit lost in the details and would like you to confirm.

Comment (claravanstaden) [reply bridges/snowbridge/pallets/ethereum-client/src/lib.rs:483]: @acatangiu that's a valid point, thanks! There is a specific test for that (`duplicate_sync_committee_updates_are_not_free`): https://github.com/paritytech/polkadot-sdk/pull/5201/files#diff-ea6025b3cac6719f750d1f0065ba30d70a16d386a5ce374e4a6b79299f48a3eeR684

Comment (acatangiu) [reply bridges/snowbridge/pallets/ethereum-client/src/lib.rs:483]: Thanks, didn't yet look at tests 🙈

File: bridges/snowbridge/pallets/ethereum-client/src/lib.rs [hunk 7]
```diff
@@ -634,11 +645,31 @@ pub mod pallet {
...
+		/// Updates are free if the update is successful and the interval between the latest
+		/// finalized header in storage and the newly imported header is large enough. All
+		/// successful sync committee updates are free.
+		pub(super) fn check_refundable(update: &Update, latest_slot: u64) -> Pays {
+			// If the sync committee was successfully updated, the update may be free.
+			if update.next_sync_committee_update.is_some() {
+				return Pays::No;
+			}
+
+			// If free headers are allowed and the latest finalized header is larger than the
```
[the remark this answers is not in the document]
Comment (claravanstaden) [reply bridges/snowbridge/pallets/ethereum-client/src/lib.rs]: Good catch, in an older implementation it was optional. Comment updated here: https://github.com/paritytech/polkadot-sdk/pull/5201/commits/ab8739cff56a96ad838020a451d9809b7f617d15#diff-adbbba03e39a047dd71745d3ed0eaf33b4e982ac58d64f273f1574816ff44940R664

File: bridges/snowbridge/pallets/ethereum-client/src/tests.rs (hunk 3 above)
Comment (claravanstaden) [reply bridges/snowbridge/pallets/ethereum-client/src/tests.rs]: Agreed, updated here: https://github.com/paritytech/polkadot-sdk/pull/5201/commits/ab8739cff56a96ad838020a451d9809b7f617d15#diff-ea6025b3cac6719f750d1f0065ba30d70a16d386a5ce374e4a6b79299f48a3eeR136

File: cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/bridge_to_ethereum_config.rs (hunk 4 above)
Comment (claravanstaden) [reply cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/bridge_to_ethereum_config.rs]: https://github.com/paritytech/polkadot-sdk/pull/5201/commits/ab8739cff56a96ad838020a451d9809b7f617d15#diff-f639f7c31fb8e9ee5ce9a5b81b1eb1eba5c9a37d05d1fa425d694cf7eec957ecR168

File: prdoc/pr_5201.prdoc (hunk 5 above)
Comment (claravanstaden) [reply prdoc/pr_5201.prdoc]: https://github.com/paritytech/polkadot-sdk/pull/5201/commits/ab8739cff56a96ad838020a451d9809b7f617d15#diff-cfa518fee4956eaf8afc94f3fad7a131b1c1474c2be560798ba1c5cec04f57dcR10

File: bridges/snowbridge/pallets/ethereum-client/src/lib.rs (hunk 6 above)
Comment (acatangiu) [reply bridges/snowbridge/pallets/ethereum-client/src/lib.rs:483]: The test tries to submit the exact same proof twice, which will fail the second time because the attested header in the proof is no longer newer than the stored one on-chain.

But what about the following scenario:
- Relayer submits valid update for slot `N` in period `P` with sync committee `S` provided in `Some(next_sync_committee_update)`
- Relayer then provides update for slot `N+1` in period `P` with sync committee `S` provided in `Some(next_sync_committee_update)`?

Basically try to "farm" the `Some(next_sync_committee_update)` tx fee freebie, to spam update one slot at a time.

**Review by acatangiu — APPROVED** [review]
Approving this like so, Clara to follow-up with requested test in another PR.

**Comment by bkchr** [comment]
@claravanstaden please merge master to make this mergable.

**Comment by claravanstaden** [comment]
> @claravanstaden please merge master to make this mergable.

On it!

File: bridges/snowbridge/pallets/ethereum-client/src/lib.rs (hunk 6 above)
Comment (acatangiu) [reply bridges/snowbridge/pallets/ethereum-client/src/lib.rs:483]: The above is a valid issue, and is being fixed in https://github.com/Snowfork/polkadot-sdk/pull/172
